File: cmd/admin/roles.py
import difflib
import logging
import re

import discord
from discord.ext import commands

from utils.core.embeds import make_embed
from utils.core.emojis import EMOJIS
from utils.logging.mod_log import send_mod_log
from utils.permissions.base_admin import BaseAdminCog

logger = logging.getLogger("bot")


class Roles(BaseAdminCog):
    """Cog responsible for managing server roles and role assignments."""

    def __init__(self, bot: commands.Bot) -> None:
        self.bot = bot

    async def _cleanup(self, ctx: commands.Context) -> None:
        try:
            await ctx.message.delete()
        except (discord.Forbidden, discord.HTTPException):
            pass

    def _find_role(self, guild: discord.Guild,
                   query: str) -> discord.Role | None:
        match = re.search(r"<@&?(\d+)>", query) or re.search(r"^(\d+)$", query)
        if match:
            return guild.get_role(int(match.group(1)))

        cleaned_query = query.lower().strip()
        roles_map = {r.name.lower(): r for r in guild.roles}

        if cleaned_query in roles_map:
            return roles_map[cleaned_query]

        substring_matches = [
            r for name, r in roles_map.items() if cleaned_query in name
        ]
        if substring_matches:
            substring_matches.sort(key=lambda r: len(r.name))
            return substring_matches[0]

        matches = difflib.get_close_matches(cleaned_query,
                                            list(roles_map.keys()),
                                            n=1,
                                            cutoff=0.5)
        return roles_map[matches[0]] if matches else None

    async def _prefix_validate(self, ctx: commands.Context,
                               target: discord.Member,
                               role: discord.Role) -> bool:
        guild = ctx.guild
        if guild is None or not isinstance(ctx.author, discord.Member):
            return False

        author = ctx.author

        if not (author.id == guild.owner_id
                or author.guild_permissions.manage_roles):
            await ctx.send(
                embed=make_embed(
                    title="Access Denied",
                    description="Requires `Manage Roles` permission.",
                    level="ERROR",
                ),
                delete_after=5,
            )
            return False

        if author.id != guild.owner_id:
            if author.top_role.position <= target.top_role.position:
                await ctx.send(
                    embed=make_embed(
                        title="Hierarchy Error",
                        description=
                        ("Your highest role must be higher than the target's highest role."
                         ),
                        level="ERROR",
                    ),
                    delete_after=5,
                )
                return False

            if author.top_role.position <= role.position:
                await ctx.send(
                    embed=make_embed(
                        title="Hierarchy Error",
                        description=
                        ("Your highest role must be higher than the role you are managing."
                         ),
                        level="ERROR",
                    ),
                    delete_after=5,
                )
                return False

        if (guild.me.top_role.position <= target.top_role.position
                or guild.me.top_role.position <= role.position):
            await ctx.send(
                embed=make_embed(
                    title="Hierarchy Error",
                    description=
                    ("I cannot manage this target or role due to my position in the role hierarchy."
                     ),
                    level="ERROR",
                ),
                delete_after=5,
            )
            return False

        return True

    @commands.group(name="role", invoke_without_command=True)
    @commands.guild_only()
    async def role(self, ctx: commands.Context) -> None:
        prefix = ctx.clean_prefix
        warning_emoji = EMOJIS.get("warning") or "⚠️"
        await ctx.send(
            embed=make_embed(
                title="Role Management",
                description=(f"{warning_emoji} Usage:\n"
                             f"`{prefix}role add <user> <role>`\n"
                             f"`{prefix}role remove <user> <role>`"),
                level="WARNING",
            ),
            delete_after=10,
        )
        await self._cleanup(ctx)

    @role.command(name="add", aliases=["give", "a"])
    @commands.guild_only()
    async def add(self, ctx: commands.Context, member: discord.Member, *,
                  role_query: str) -> None:
        guild = ctx.guild
        if guild is None or not isinstance(ctx.author, discord.Member):
            return

        role = self._find_role(guild, role_query)
        if not role:
            await ctx.send(
                embed=make_embed(
                    title="Not Found",
                    description=f"Role `{role_query}` could not be found.",
                    level="ERROR",
                ),
                delete_after=5,
            )
            await self._cleanup(ctx)
            return

        if not await self._prefix_validate(ctx, member, role):
            await self._cleanup(ctx)
            return

        if role in member.roles:
            await ctx.send(
                embed=make_embed(
                    title="No Change",
                    description=f"{member.display_name} already has this role.",
                    level="WARNING",
                ),
                delete_after=5,
            )
        else:
            try:
                await member.add_roles(role, reason=f"Managed by {ctx.author}")
                success_emoji = EMOJIS.get("success") or "✅"
                await ctx.send(
                    embed=make_embed(
                        title="Success",
                        description=
                        f"{success_emoji} Added {role.mention} to {member.mention}.",
                        level="SUCCESS",
                    ),
                    delete_after=5,
                )

                try:
                    await send_mod_log(
                        guild=guild,
                        category="MODERATION",
                        title="Role Added",
                        description=
                        f"{ctx.author.mention} gave {role.mention} to {member.mention}",
                        level="INFO",
                        actor=ctx.author,
                        target=member,
                        extra_fields={"Role": role.name},
                    )
                except Exception:
                    logger.exception(
                        "Failed to send role addition moderation log")

            except discord.HTTPException:
                await ctx.send(embed=make_embed(
                    title="Execution Error",
                    description="Failed to apply role modifications.",
                    level="ERROR"),
                               delete_after=5)

        await self._cleanup(ctx)

    @role.command(name="remove", aliases=["take", "r"])
    @commands.guild_only()
    async def remove(self, ctx: commands.Context, member: discord.Member, *,
                     role_query: str) -> None:
        guild = ctx.guild
        if guild is None or not isinstance(ctx.author, discord.Member):
            return

        role = self._find_role(guild, role_query)
        if not role:
            await ctx.send(embed=make_embed(
                title="Not Found",
                description=f"Role `{role_query}` could not be found.",
                level="ERROR"),
                           delete_after=5)
            await self._cleanup(ctx)
            return

        if not await self._prefix_validate(ctx, member, role):
            await self._cleanup(ctx)
            return

        if role not in member.roles:
            await ctx.send(embed=make_embed(
                title="No Change",
                description=f"{member.display_name} does not have this role.",
                level="WARNING"),
                           delete_after=5)
        else:
            try:
                await member.remove_roles(role,
                                          reason=f"Managed by {ctx.author}")
                success_emoji = EMOJIS.get("success") or "✅"
                await ctx.send(embed=make_embed(
                    title="Success",
                    description=
                    f"{success_emoji} Removed {role.mention} from {member.mention}.",
                    level="SUCCESS"),
                               delete_after=5)

                try:
                    await send_mod_log(
                        guild=guild,
                        category="MODERATION",
                        title="Role Removed",
                        description=
                        f"{ctx.author.mention} removed {role.mention} from {member.mention}",
                        level="INFO",
                        actor=ctx.author,
                        target=member,
                        extra_fields={"Role": role.name})
                except Exception:
                    logger.exception(
                        "Failed to send role removal moderation log")

            except discord.HTTPException:
                await ctx.send(embed=make_embed(
                    title="Execution Error",
                    description="Failed to remove role modifications.",
                    level="ERROR"),
                               delete_after=5)

        await self._cleanup(ctx)


async def setup(bot: commands.Bot) -> None:
    await bot.add_cog(Roles(bot))

File: src/commands/admin/roles.js
import { PermissionFlagsBits, SlashCommandBuilder } from "discord.js";
import { createCommand } from "../../core/command.js";
import { makeEmbed } from "../../core/embeds.js";
import { EMOJIS } from "../../core/emojis.js";
import { sendModLog } from "../../utils/modLog.js";

const slashBuilder = new SlashCommandBuilder()
  .setName("role")
  .setDescription("Assign or remove a role from a member")
  .addSubcommand((sub) =>
    sub
      .setName("add")
      .setDescription("Give a role to a member")
      .addUserOption((opt) => opt.setName("user").setDescription("The target member").setRequired(true))
      .addRoleOption((opt) => opt.setName("role").setDescription("The role to give").setRequired(true))
  )
  .addSubcommand((sub) =>
    sub
      .setName("remove")
      .setDescription("Remove a role from a member")
      .addUserOption((opt) => opt.setName("user").setDescription("The target member").setRequired(true))
      .addRoleOption((opt) => opt.setName("role").setDescription("The role to remove").setRequired(true))
  );

export default createCommand({
  name: "role",
  description: "Assign or remove a role from a member",
  category: "Admin",
  aliases: ["roles", "giverole", "removerole"],
  modOnly: true,
  requiredPermission: PermissionFlagsBits.ManageRoles,
  slashBuilder,

  async execute(ctx) {
    const { guild, member: actorMember, user } = ctx;
    if (!guild) return;

    // ── Resolve subcommand ──────────────────────────────────────────────────
    // Slash: ctx.subcommand = "add" | "remove"
    // Prefix: ts role add @user @role  →  _args[0]="add", _args[1]="<@id>", _args[2]="<@&id>"
    //         ts giverole @user @role  →  (alias) treated as "add"
    //         ts removerole @user @role→  (alias) treated as "remove"
    const invokedAs = ctx.message?.content
      ?.trim()
      .split(/\s+/)[0]
      ?.toLowerCase()
      .replace(/^[^\w]*/, "");

    let sub = ctx.subcommand;
    if (!sub) {
      const firstArg = ctx.options._args?.[0]?.toLowerCase();
      if (firstArg === "add" || firstArg === "remove") {
        sub = firstArg;
      } else if (invokedAs === "removerole") {
        sub = "remove";
      } else {
        // default to "add" for giverole / bare "role" with no sub
        sub = "add";
      }
    }

    // ── Resolve target user ────────────────────────────────────────────────
    let targetUserId = ctx.options.user; // slash: resolved user ID
    if (!targetUserId) {
      // prefix: skip first arg if it was the subcommand keyword
      const argOffset = (ctx.options._args?.[0]?.toLowerCase() === "add" || ctx.options._args?.[0]?.toLowerCase() === "remove") ? 1 : 0;
      const rawUser = ctx.options._args?.[argOffset] || "";
      targetUserId = rawUser.replace(/[<@!>]/g, "");
    }

    // ── Resolve role ───────────────────────────────────────────────────────
    let roleId = ctx.options.role; // slash: resolved role ID
    if (!roleId) {
      const argOffset = (ctx.options._args?.[0]?.toLowerCase() === "add" || ctx.options._args?.[0]?.toLowerCase() === "remove") ? 2 : 1;
      const rawRole = ctx.options._args?.[argOffset] || "";
      roleId = rawRole.replace(/[<@&>]/g, "");
    }

    // ── Validation ─────────────────────────────────────────────────────────
    if (!targetUserId || !roleId) {
      const prefix = ctx.client?.prefix || "ts";
      return await ctx.reply({
        embeds: [
          makeEmbed({
            title: "Usage",
            description:
              `**Slash:**\n\`/role add <user> <role>\`\n\`/role remove <user> <role>\`\n\n` +
              `**Prefix:**\n\`${prefix}role add @user @role\`\n\`${prefix}giverole @user @role\`\n\`${prefix}removerole @user @role\``,
            level: "INFO",
          }),
        ],
        ephemeral: true,
      });
    }

    const targetMember = await guild.members.fetch(targetUserId).catch(() => null);
    const role = guild.roles.cache.get(roleId) || guild.roles.cache.find((r) => r.id === roleId);

    if (!targetMember) {
      return await ctx.reply({
        embeds: [makeEmbed({ title: "Not Found", description: `${EMOJIS.get("fail") || "❌"} Member not found in this server.`, level: "ERROR" })],
        ephemeral: true,
      });
    }

    if (!role) {
      return await ctx.reply({
        embeds: [makeEmbed({ title: "Not Found", description: `${EMOJIS.get("fail") || "❌"} Role not found. Make sure you @mention the role.`, level: "ERROR" })],
        ephemeral: true,
      });
    }

    const botMember = guild.members.me;

    // Bot hierarchy check
    if (role.position >= botMember.roles.highest.position) {
      return await ctx.reply({
        embeds: [
          makeEmbed({
            title: "Hierarchy Error",
            description: `${EMOJIS.get("fail") || "❌"} I cannot manage ${role} — it is at or above my highest role.`,
            level: "ERROR",
          }),
        ],
        ephemeral: true,
      });
    }

    // Actor hierarchy check
    if (actorMember && role.position >= actorMember.roles.highest.position && guild.ownerId !== user.id) {
      return await ctx.reply({
        embeds: [
          makeEmbed({
            title: "Permission Denied",
            description: `${EMOJIS.get("fail") || "❌"} You cannot manage ${role} — it is at or above your highest role.`,
            level: "ERROR",
          }),
        ],
        ephemeral: true,
      });
    }

    // ── Execute ────────────────────────────────────────────────────────────
    if (sub === "add") {
      if (targetMember.roles.cache.has(role.id)) {
        return await ctx.reply({
          embeds: [
            makeEmbed({
              title: "Already Assigned",
              description: `${EMOJIS.get("warning") || "⚠️"} ${targetMember} already has the ${role} role.`,
              level: "WARNING",
            }),
          ],
          ephemeral: true,
        });
      }

      await targetMember.roles.add(role, `Role assigned by ${user.tag || user.username}`).catch(() => {});

      await sendModLog({
        guild,
        title: "Role Added",
        description: `${role} given to ${targetMember} by ${user}.`,
        level: "INFO",
        actor: user,
        extraFields: { Role: role.name, Member: targetMember.user.tag || targetMember.user.username },
      }).catch(() => {});

      return await ctx.reply({
        embeds: [
          makeEmbed({
            author: { name: "Role Management", iconURL: guild.iconURL?.({ dynamic: true }) || undefined },
            title: "Role Added",
            description:
              `${EMOJIS.get("success") || "✅"} Successfully gave ${role} to ${targetMember}.\n\n` +
              `• **Member:** ${targetMember} (\`${targetMember.user.tag || targetMember.user.username}\`)\n` +
              `• **Role:** ${role} (\`${role.name}\`)\n` +
              `• **Moderator:** ${user}`,
            level: "SUCCESS",
            headerDivider: false,
          }),
        ],
      });
    }

    if (sub === "remove") {
      if (!targetMember.roles.cache.has(role.id)) {
        return await ctx.reply({
          embeds: [
            makeEmbed({
              title: "Not Assigned",
              description: `${EMOJIS.get("warning") || "⚠️"} ${targetMember} does not have the ${role} role.`,
              level: "WARNING",
            }),
          ],
          ephemeral: true,
        });
      }

      await targetMember.roles.remove(role, `Role removed by ${user.tag || user.username}`).catch(() => {});

      await sendModLog({
        guild,
        title: "Role Removed",
        description: `${role} removed from ${targetMember} by ${user}.`,
        level: "INFO",
        actor: user,
        extraFields: { Role: role.name, Member: targetMember.user.tag || targetMember.user.username },
      }).catch(() => {});

      return await ctx.reply({
        embeds: [
          makeEmbed({
            author: { name: "Role Management", iconURL: guild.iconURL?.({ dynamic: true }) || undefined },
            title: "Role Removed",
            description:
              `${EMOJIS.get("success") || "✅"} Successfully removed ${role} from ${targetMember}.\n\n` +
              `• **Member:** ${targetMember} (\`${targetMember.user.tag || targetMember.user.username}\`)\n` +
              `• **Role:** ${role} (\`${role.name}\`)\n` +
              `• **Moderator:** ${user}`,
            level: "SUCCESS",
            headerDivider: false,
          }),
        ],
      });
    }
  },
});
